src/model/CoilWaterHeatingDesuperheater.cpp

```diff
@@ -413,20 +377,64 @@ namespace model {
+                                          << " <= 0.3 when Heating Source Object Type = " << heatingSourceIddObjectType.valueDescription() << ".");
+            return false;
+          }
+        }
+      }
+
+      bool result = setDouble(OS_Coil_WaterHeating_DesuperheaterFields::RatedHeatReclaimRecoveryEfficiency, ratedHeatReclaimRecoveryEfficiency);
+      return result;
+    }
+
+    void CoilWaterHeatingDesuperheater_Impl::resetRatedHeatReclaimRecoveryEfficiency() {
+      bool result = setString(OS_Coil_WaterHeating_DesuperheaterFields::RatedHeatReclaimRecoveryEfficiency, "");
+      OS_ASSERT(result);
+    }
 
-    bool CoilWaterHeatingDesuperheater_Impl::setHeatingSource(const boost::optional<ModelObject>& heatingSource) {
-      bool result(false);
-      if (heatingSource) {
-        result = setPointer(OS_Coil_WaterHeating_DesuperheaterFields::HeatingSourceName, heatingSource.get().handle());
-      } else {
-        resetHeatingSource();
-        result = true;
+    bool CoilWaterHeatingDesuperheater_Impl::setHeatingSource(const ModelObject& heatingSource) {
+      bool result = setPointer(OS_Coil_WaterHeating_DesuperheaterFields::HeatingSourceName, heatingSource.handle());
+
+      if (boost::optional<double> ratedHeatReclaimRecoveryEfficiency_ = this->ratedHeatReclaimRecoveryEfficiency()) {
+        auto heatingSourceIddObjectType = heatingSource.iddObjectType();
+        if ((heatingSourceIddObjectType == IddObjectType::OS_Refrigeration_Condenser_AirCooled)
+            || (heatingSourceIddObjectType == IddObjectType::OS_Refrigeration_Condenser_EvaporativeCooled)
+            || (heatingSourceIddObjectType == IddObjectType::OS_Refrigeration_Condenser_WaterCooled)) {
+          if (ratedHeatReclaimRecoveryEfficiency_.get() > 0.9) {
+            LOG(Error, briefDescription() << ": Rated Heat Reclaim Recovery Efficiency must be"
+                                          << " <= 0.9 when Heating Source Object Type = " << heatingSourceIddObjectType.valueDescription()
+                                          << ". Resetting it to the default of 0.8.");
+            // setDouble to avoid a double check
+            // setRatedHeatReclaimRecoveryEfficiency(0.8);
+            setDouble(OS_Coil_WaterHeating_DesuperheaterFields::RatedHeatReclaimRecoveryEfficiency, 0.8);
+          }
+        } else {
+          if (ratedHeatReclaimRecoveryEfficiency_.get() > 0.3) {
+            LOG(Error, briefDescription() << ": Rated Heat Reclaim Recovery Efficiency must be"
+                                          << " <= 0.3 when Heating Source Object Type = " << heatingSourceIddObjectType.valueDescription()
+                                          << ". Resetting it to the default of 0.25.");
+            // setRatedHeatReclaimRecoveryEfficiency(0.25);
+            setDouble(OS_Coil_WaterHeating_DesuperheaterFields::RatedHeatReclaimRecoveryEfficiency, 0.25);
+          }
+        }
```

Comment:
When there is already a Rated Heat Reclaim Efficiency assigned, when you use setHeatingSource => check that it matches, and if not issue an error message and reset to the E+ default (0.25 / 0.8)